### PR TITLE
[DependencyInjection] Add type-hints to interfaces and implementations.

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #32179 
| License       | MIT
| Doc PR        | N/A

Part II for DependencyInjection, concentrating on the main interfaces, `ContainerBuiler` and related classes.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/ContainerInterface.php
Comment (derrabus) [inline src/Symfony/Component/DependencyInjection/ContainerInterface.php]: Note: We cannot add `string $id` because we would violate the PSR interface otherwise.

File: src/Symfony/Component/DependencyInjection/Definition.php
Comment (derrabus) [inline src/Symfony/Component/DependencyInjection/Definition.php]: That one gave me quite a headache. I will open a PR on 3.4 to adjust the type-hint.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Should be reverted as it does not look related to the purpose of this PR, is it?

File: src/Symfony/Component/DependencyInjection/Definition.php
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Definition.php]: Should be reverted (same for other similar changes) as this "use case" should be deprecated in 4.4 instead.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php
Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: This class generates implementations of the `ContainerInterface` that has been changed with this PR. This is why I've adjusted the code template above to add the new type declarations and remove the now-obsolete type cast.

This `preg_replace` call removes an `if` block from the generated code under certain circumstances. Because I've removed the type cast, this operation would leave an empty line at the beginning of the method, like this:

```php
public function hasParameter(string $name)
{

    return isset($this->parameters[$name]) || isset($this->loadedDynamicParameters[$name]) || array_key_exists($name, $this->parameters);
}
```

I've adjusted the regular expression, so that this extra new line is also removed in case its there.

File: src/Symfony/Component/DependencyInjection/Definition.php
Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/Definition.php]: Yes. Let's put this on hold until the deprecation is merged.

Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/Definition.php]: #32390

File: src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/ShareTrait.php
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/ShareTrait.php]: the param phpdoc can be reoved

File: src/Symfony/Component/DependencyInjection/Definition.php
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Definition.php]: param phpdoc can be removed

**Comment by Tobion** [comment]
There are still several uneccessary string casts left, e.g.

- https://github.com/symfony/symfony/blob/492df5e21abeef6750f34d3e63a4db63f6a915a0/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L507
-https://github.com/symfony/symfony/blob/492df5e21abeef6750f34d3e63a4db63f6a915a0/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L859
- https://github.com/symfony/symfony/blob/492df5e21abeef6750f34d3e63a4db63f6a915a0/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L960

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: `setAutowired` has been forgotten to add the type

Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: indeed

**Comment by Tobion** [comment]
There seems to be something wrong in the failing tests testBindScalarValueToControllerArgument and testRegistersTraceableBusesToCollector

**Comment by derrabus** [comment]
I'll look into it.

**Comment by derrabus** [comment]
All right, so the failing test assumes that a service can be a scalar value.

https://github.com/symfony/symfony/blob/6811aaa8e087665769dd66c69fb3109d002196ef/src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterControllerArgumentLocatorsPassTest.php#L348

All docblocks on the DI classes and interfaces however promise me that services are always objects.

https://github.com/symfony/symfony/blob/6811aaa8e087665769dd66c69fb3109d002196ef/src/Symfony/Component/DependencyInjection/ContainerInterface.php#L37

https://github.com/symfony/symfony/blob/6811aaa8e087665769dd66c69fb3109d002196ef/src/Symfony/Component/DependencyInjection/ContainerInterface.php#L47

https://github.com/symfony/symfony/blob/6811aaa8e087665769dd66c69fb3109d002196ef/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L1641

I trusted these docblocks and used `object` type-hints for services and those make the test mentioned by @Tobion fail. Interestingly enough, all tests of the DI component itself are green with the object type-hints. The failing test I linked above comes from the HttpKernel component.

I'm unsure if scalar services should be a thing, but they're possible at the moment, at least to a certain degree. In order to move forward here, I'll remove the `object` type-hints for now, so that we can decide afterwards if we either want to deprecate scalar services or update the docblocks.

**Comment by Tobion** [comment]
According to PSR-11 it could also be a scalar value.

> use of a container to obtain objects and parameters

But that seems strange and not the original contract of the symfony container. Also parameters are not fetched individually from the container but through a ParameterBag in Symfony which makes more sense to me.

**Comment by derrabus** [comment]
The remaining failing tests are unrelated to my changes.

Status: Needs Review

**Comment by derrabus** [comment]
Follow-up issue: #32411

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @derrabus.

**Pull request merged**
